go/pkg/cas/upload.go

```diff
@@ -120,6 +149,102 @@ func (in *UploadInput) DigestsComputed() <-chan struct{} {
 	return in.ensureDigestsComputedInited()
 }
 
+var oneDot = []string{"."}
+
+// init initializes internal fields.
+func (in *UploadInput) init(u *uploader) error {
+	in.u = u
+
+	if !filepath.IsAbs(in.Path) {
+		return errors.Errorf("%q is not absolute", in.Path)
+	}
+	in.cleanPath = filepath.Clean(in.Path)
+
+	// Do not use os.Stat() here. We want to know if it is a symlink.
+	var err error
+	if in.pathInfo, err = os.Lstat(in.cleanPath); err != nil {
+		return errors.WithStack(err)
+	}
+
+	// Process the allowlist.
+	in.tree = make(map[string]*digested, 1+len(in.Allowlist))
+	switch {
+	case len(in.Allowlist) == 0:
+		in.cleanAllowlist = oneDot
+
+	case in.pathInfo.Mode().IsRegular():
+		return errors.Errorf("the Allowlist is not supported for regular files")
+
+	default:
+		in.cleanAllowlist = make([]string, len(in.Allowlist))
+		for i, subPath := range in.Allowlist {
+			if filepath.IsAbs(subPath) {
+				return errors.Errorf("the allowlisted path %q is not relative", subPath)
+			}
+
+			cleanSubPath := filepath.Clean(subPath)
+			if cleanSubPath == ".." || strings.HasPrefix(cleanSubPath, parentDirPrefix) {
+				return errors.Errorf("the allowlisted path %q contains `..`", subPath)
+			}
+			in.cleanAllowlist[i] = cleanSubPath
+		}
+	}
+	return nil
+}
+
+// partialMerkleTree ensures that for each node in in.tree, not included by any
+// other node, all its ancestors are also present in the tree. For example, if
+// the tree contains only "foo/bar" and "foo/baz", then partialMerkleTree adds
+// "foo" and ".". The latter is the root.
+//
+// All tree keys must be clean relative paths.
+// Returns prepared *uploadItems that represent the ancestors that were added to
+// the tree.
+func (in *UploadInput) partialMerkleTree() (added []*uploadItem) {
+	// Establish parent->child edges.
+	children := map[string]map[string]struct{}{}
+	for relPath := range in.tree {
+		for relPath != "." {
+			parent := dirNameRelFast(relPath)
+			if childSet, ok := children[parent]; ok {
+				childSet[relPath] = struct{}{}
+			} else {
+				children[parent] = map[string]struct{}{relPath: {}}
+			}
+			relPath = parent
+		}
+	}
+
+	// Add the missing ancestors by traversing in post-order.
+	var dfs func(relPath string) proto.Message
+	dfs = func(relPath string) proto.Message {
```

Comment:
Maybe just return it, and add "dfs" to the comment? (post-order should be explicit enough, but dfs may be easier to ring bells to the reader). No strong feelings.
```suggestion
	return func(relPath string) proto.Message { ... } (".")
```

Reply:
this is a recursive func. Without `dfs`, it cannot call itself.
If it wasn't recursive, we wouldn't need the func to begin with.

Reply:
also, the return value of dfs() is not what partialMerkleTree returns. The latter returns `added`, which is captured in the func.